http: fix parsing of binary upgrade response body

bnoordhuis opened the pull request:
Fix a bug where a connection upgrade response with a Transfer-Encoding
header and a body whose first byte is > 127 causes said byte to be
dropped on the floor when passing the remainder of the message to
the 'upgrade' event listeners.

Fixes: https://github.com/nodejs/node/issues/17789
CI: https://ci.nodejs.org/job/node-test-commit/15003/

bnoordhuis commented on lib/_http_client.js:
```diff
@@ -471,29 +470,22 @@ function parserOnIncomingClient(res, shouldKeepAlive) {
     // sent a double response.
     socket.destroy();
-    return;
+    return 0;  // No special treatment.
```
Before anyone asks - yes, these could be constants, please suggest good names. :-)

(They don't have names in http_parser either though.)

cjihrig replied:
Just made these up on the spot. Feel free to ignore. How about `kSkipBodyAsUpgrade`, `kSkipBodyNoUpgrade`... and I don't know for the 0 case.

bnoordhuis commented on lib/_http_client.js:
```diff
@@ -471,29 +470,22 @@ function parserOnIncomingClient(res, shouldKeepAlive) {
-  // to the content-length of the entity-body had the request
-  // been a GET.
-  var isHeadResponse = req.method === 'HEAD';
-  debug('AGENT isHeadResponse', isHeadResponse);
```
Removed because not very useful and slightly misleading - this is not inside the agent.

bnoordhuis commented on lib/_http_server.js:
```diff
@@ -616,7 +616,7 @@ function parserOnIncoming(server, socket, state, req, keepAlive) {
   } else {
     server.emit('request', req, res);
   }
-  return false; // Not a HEAD response. (Not even a response!)
```
Bogus comment, removed.

bnoordhuis commented on lib/_http_common.js:
```diff
@@ -106,19 +106,10 @@ function parserOnHeadersComplete(versionMajor, versionMinor, headers, method,
-  if (!upgrade) {
-    // For upgraded connections and CONNECT method request, we'll emit this
-    // after parser.execute so that we can capture the first part of the new
-    // protocol.
```
Slightly wrong comment, removed.

richardlau commented on test/parallel/test-http-upgrade-binary.js:
```diff
@@ -0,0 +1,28 @@
+// https://github.com/nodejs/node/issues/17789 - a connection upgrade response
+// that has a Transfer-Encoding header and a body whose first byte is > 127
+// triggers a bug where said byte is skipped over.
+net.createServer(function(conn) {
```
This callback and the one passed to `listen` could also be wrapped in `common.mustCall`.

maclover7 commented:
ping @bnoordhuis

joyeecheung commented:
Is this ready to land? Or do you want to fix the nits before this PR gets landed? @bnoordhuis

bnoordhuis commented:
Rebased + new CI: https://ci.nodejs.org/job/node-test-pull-request/12627/

I've incorporated the comments about the test.  I decided to leave the constants unnamed.

BridgeAR commented:
Landed in de4600ea2bf2eb5a1367825dc84ea7561b0c3529

MylesBorins commented:
Should this be backported to LTS? It lands cleanly on 8.x but will need a manual backport to v6.x

bnoordhuis commented:
Yes, should be back-ported.  I'll look into the v6.x back-port.

jeroenvollenbrock commented:
@gibfahn Is there any reason this one was not included in 8.10.0? I can confirm the bug persists in node 8.11.1 :)

MylesBorins commented:
@jeroenvollenbrock not sure why it was missed... I've landed it on staging

@bnoordhuis did you have a chance to look into the 6.x backport?

shellscape commented:
@lpinca @MylesBorins it looks like Node 10.14.1 has a regression and this problem has snuck back into the release. This gist can consistently reproduce the error: https://gist.github.com/shellscape/2020916f92be5d61f4d411fb6d2bce61

Note that stepping back down to any 10.x version lower than 10.14.0 does not reproduce the error.

lpinca commented:
@shellscape none of these commits https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md#commits seem related.

shellscape commented:
@lpinca indeed, which baffled me. we've spun up brand new containers to verify the issue was isolated to 10.14.x. same containers from scratch with 10.13.x are completely fine.

MylesBorins commented:
I'll dig into the repro and see what's going on

shellscape commented:
@MylesBorins much appreciated. if you end up needing that in a separate repository, say the word and it shall be done.

BridgeAR commented:
@shellscape would you be so kind and open a new issue about this with as much information and possible? That way it's easier to track the progress :)

shellscape commented:
@BridgeAR I'll do my best, please don't nuke it right away if there's something lacking, happy to clean it up to spec. It'll be my first Node org issue.

BridgeAR commented:
Don't worry, we normally only close an issue if they it's resolved, not a core issue or in the wrong repository.

As @lpinca pointed out, it's best to try to reduce the test case further by removing any third party modules. That way looking into it becomes much easier.

shellscape commented:
@BridgeAR https://github.com/nodejs/node/issues/24958. Trying to reproduce this with core modules is a bit beyond my range of ability. Gave it a shot, but there's a lot of knowledge needed to reproduce the same conditions that I apparently don't have as yet.
